### PR TITLE
[core] Add PREPARED state for placement groups.

### PR DESCRIPTION
**Pull request by rynewang**
Ray Placement Groups use 2pc protocol to reserve resources. However we don't track PREPARED state; we only had PENDING (before any prepare) and CREATED (after all committed). This means when the GCS restarts, it does not know any Commits ever happened. Indeed, it instructs all raylets to return all bundles prepared or committed.

If the GCS restart happens when a PG has finished all Prepare and is doing Commit, the already-Committed bundles may have scheduled actors/tasks, and those actors/tasks may be killed by the GCS restart, making the GCS restart user-visible.

This PR introduces a PREPARED state between PENDING and CREATED. Rules:

- (New!) On a PG received all Prepare replies successfully: state change PENDING -> PREPARED, and persist to Redis. This also persists all bundle locations (node IDs).
- After the persist: send all Commit calls.
- On a PG received all Commit replies successfully: state change PREPARED -> CREATED, and persist to Redis.
- On GCS Restart, if a PG is PENDING: return any bundles, reschedule from zero.
- (New!) On GCS Restart, if a PG is PREPARED, continue to send all Commit calls.

In order to do this cleanly, did a bunch of refactoring. The idea is to keep all interface changes to the `GcsPlacementGroupScheduler` but not to `GcsPlacementGroupManager`. Specifically:

- `GcsPlacementGroupScheduler::ScheduleUnplacedBundles` now accepts an aggregated arg struct `SchedulePgRequest`.
- `GcsPlacementGroupScheduler::Initialize` accepts one more vector of arg struct `SchedulePreparedPgRequest`, which is the regular `SchedulePgRequest` plus the prepared bundle information.
- `GcsPlacementGroupManager::Initialize` now distinguishes PREPARED PGs by creating requests and send them to the Scheduler. They are also kept in `used_bundles` to restrain from being removed (<- actual fix).
- Unit test for the original issue. Failing on master and passing on this PR.
- Quality of life changes: `RAY_LOG().WithField(PlacementGroupID)`.

### CODE REVIEW

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc
Comment (jjyao) [inline src/ray/gcs/gcs_server/gcs_placement_group_scheduler.cc:819]: No need to pass in `prepared_bundles` since you can get it from `placement_group`?

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h
Comment (jjyao) [inline src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h:60]: Seems we don't need `SchedulePreparedPgRequest` as we can get bundles from `SchedulePgRequest.placement_group`

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h
Comment (jjyao) [inline src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h:68]: ?

**Comment by jjyao** [comment]
Do we have tests that during recommitting the prepared bundles, some recommits failed and the PG goes to RESCHEDULING state?

**Comment by rynewang** [comment]
> Do we have tests that during recommitting the prepared bundles, some recommits failed and the PG goes to RESCHEDULING state?

updated python test.

File: python/ray/tests/test_placement_group_failover.py
Comment (jjyao) [inline python/ray/tests/test_placement_group_failover.py:266]: can we get rid of time.sleep using `wait_for_condition` etc?

**Review by jjyao — APPROVED** [review]
lgtm.

In the future, we can refactor the code to be more reconciliation style.

Please wait for @rkooo567's approval before merging.

**Comment by rynewang** [comment]
@rkooo567 can you take a look on this one? if you are ok with it pls merge

**Comment by rkooo567** [comment]
let me try this by today. If it is urgent, please merge it without my approval

File: src/ray/gcs/gcs_server/gcs_placement_group_manager.cc
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_placement_group_manager.cc:31]: can you use constant for 1000000

**Review by rkooo567 — COMMENTED** [review]
WIP review. I think the motivation makes a lot of sense. One thing I'd like to be careful is to handle edge cases in removing pg (I will need time to review this part of code one more time..). In the meantime, can you run a release test? 

- stress_test_placement_group
- pg_autoscaling_regression_test
- placement_group_performance_test

I think if stress_test_placement_group passes, I don't worry much about handling removal of pg after introducing a new state

**Review by rkooo567 — COMMENTED** [review]
Looks pretty good! Just a couple comments on unit tests; AFter that, let's merge

- Can you make sure we handle removing placement group is handled properly (add unit tests) if it is removed after redis.Put is called (and before it is returned)? I barely remembered we don't remove pg that's in progress to simplify the protocol, but I am not 100% sure if it is still the case. 
- CreatePrepared
- cpp tests in gcs placement group manager (if exists)?

**Comment by rynewang** [comment]
So the 3 release tests passed: https://buildkite.com/ray-project/release/builds/20250

File: src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h
Comment (rynewang) [reply src/ray/gcs/gcs_server/gcs_placement_group_scheduler.h:356]: This is still correct

**Review by rkooo567 — APPROVED** [review]
let's merge it after adding remove unit tests to cpp (remove happens after Put(PREPARED) is sent and before it is returned). I think there's no race condition now given stress tests pass but just in case

**Comment by rynewang** [comment]
Just to be safe I plan to add a "is cancelled?" check after the put (in the put callback).

**Comment by rkooo567** [comment]
that sounds good. or some kind of assert that it is not removed if we don't remove pg while it is scheduled (I think this was the case last time)

**Comment by rynewang** [comment]
For the Put-then-cancel case it should already work: the pg is committed, and after all-committed the scheduler finds it marked cancelled and cancel it. But sure I added a TOCTOU check before doing any committing, and destroy it. this should prevent any tasks/actors from starting.

**Comment by rynewang** [comment]
wydm "we don't remove pg while it is scheduled"? I think user can always cancel it at any time?

**Comment by rynewang** [comment]
@rkooo567 cpp test added, "is it cancelled?" check before starting any commit added

**Comment by rkooo567** [comment]
let's merge!

**Comment by can-anyscale** [comment]
this is broken window builds https://buildkite.com/ray-project/postmerge/builds/5995#_, i'm reverting to unblock @rynewang

**Pull request merged**
